src/extension/src/coreTemplateStudio.ts

```diff
@@ -94,144 +84,151 @@ export class CoreTemplateStudio {
     }
   }
 
-  private constructor(process: ChildProcess, port: number, url: string) {
-    this._process = process;
-    this._port = port;
-    this._url = url;
+  private constructor(processCli: ChildProcess) {
+    this._processCli = processCli;
+    this.cliEvents = new CliEventEmitter();
+    this.readStream(this._processCli);
   }
 
-  public getPort(): number {
-    return this._port;
+  // This function is a listener, in the constructor, it gets attached
+  // then it will always get triggered when a command is write to the cli and there is responses from cli, until the process gets killed
+  public async readStream(process: ChildProcess) {
+    let data = "";
+    process.stdout.on("data", (chunk) => {
+      data += chunk;
+      var responses = data.toString().split("\n");
+      for (let i = 0; i < responses.length - 1; i++) {
+        var result = JSON.parse(responses[i]);
+        this.cliEvents.emit(result['messageType'], result['content']);
+      }
+      data = responses[responses.length - 1];
+    });
+
+    process.stderr.on("data", (data) => {
+      this.cliEvents.emit("error", data.toString());
+    });
   }
 
-  public async getProjectTypes(): Promise<any> {
-    // TODO: use this in client instead of fetching directly from API server
-    const url = new URL(CONSTANTS.API.ENDPOINTS.PROJECT_TYPE, this._url);
-    return await fetch(url.href, { method: CONSTANTS.API.METHODS.GET })
-      .then((response: Response) => {
-        return response.json();
-      })
-      .catch((error: Error) => {
-        throw Error(error.toString());
+  private async awaitCliEvent(eventName: string): Promise<any> {
+    return new Promise((resolve, reject) => {
+      this.cliEvents.once(eventName, (data) => {
+        this.cliEvents.removeAllListeners();
+        resolve(data);
+      });
+      this.cliEvents.once("error", (data) => {
+        this.cliEvents.removeAllListeners();
+        reject(data);
       });
+    });
   }
 
-  public async getFrameworks(projectType: string): Promise<any> {
-    const url = new URL(CONSTANTS.API.ENDPOINTS.FRAMEWORK, this._url);
-    url.searchParams.append(
-      CONSTANTS.API.QUERY_PARAMS.PROJECT_TYPE,
-      projectType
-    );
+  public async sync(payload: ICommandPayload): Promise<any> {
+    const syncCommand = `${CONSTANTS.CLI.SYNC_COMMAND_PREFIX} -p ${payload.payload.path}\n`;
+    this._processCli.stdin.write(syncCommand);
+    this.cliEvents.on(CONSTANTS.CLI.SYNC_PROGRESS_STATE, (data) => {
+      payload.liveMessageHandler(data['status'], data['progress']);
+    });
+    return await this.awaitCliEvent(CONSTANTS.CLI.SYNC_COMPLETE_STATE);
```

Comment:
A note on what your answer was to David:
> The reason is I want promise get resolved before it gets returned. Thus I use await here. Await puts an pause in the code.

The only _pause_ that would happen here would be before resolving the promise to the caller of the method. In the case where you're returning a promise, if you `await` for it to resolve before returning it (and thus resolving it), it adds one layer more than needed to do **exactly the same thing**.

Now, this is a common mistake when writing asynchronous functions in TS/JS, which is why there is an [ESLint rule for it that is well documented](https://eslint.org/docs/rules/no-return-await). If you are still unsure about the behavior, I would recommend [reading this short article](https://jakearchibald.com/2017/await-vs-return-vs-return-await/) that explains it more thoroughly. 😊

Reply:
This is what I read from [JavaScript documents ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)about **await**: 

> The await expression causes async function execution to pause until a Promise is settled, that is fulfilled or rejected, and to resume execution of the async function after fulfillment. When resumed, the value of the await expression is that of the fulfilled Promise.

In the article you linked, it gives better insights about promise/await. I guess because I am using **return** a promise here rather than a promise, **return** has the logic to wait and resolve a promise. Thus, I could emit the **await** in my code. 

It also points out **return** a promise will resolve the defer the result, while **return await** a promise will turn its rejection into a throw statement, which will be useful if there is a try catch.

Thank you so much for sharing this resource and will update my PR accordingly :)